xapian-applications/omega/handler_libarchive.cc

```diff
@@ -31,6 +34,215 @@
+		metadata.resize(size);
+		parse_metadata(metadata, title, keywords, author);
+	    }
+	}
+    }
+    OpenDocParser parser;
+    parser.parse(content + styles);
+    dump = parser.dump;
+
+    return true;
+}
+
+static bool
+extract_open_xml(struct archive* archive_obj,
+		 string& tail,
+		 string& dump,
+		 string& title,
+		 string& keywords,
+		 string& author,
+		 string& error)
+{
+    size_t total;
+    ssize_t size;
+    struct archive_entry* entry;
+    string content;
+    bool msxml = false;
+
+    if (startswith(tail, "wordprocessingml.")) {
+	msxml = true;
+	while (archive_read_next_header(archive_obj, &entry) == ARCHIVE_OK) {
+	    string pathname = archive_entry_pathname(entry);
+	    if (pathname == "word/document.xml") {
+		auto i = content.size();
+		total = archive_entry_size(entry);
+		content.resize(i + total);
+		size = archive_read_data(archive_obj, &content[i], total);
+
+		if (size <= 0) {
+		    error = "Libarchive was not able to extract data "
+			    "from word/document.xml";
+		    return false;
+		}
+		content.resize(i + size);
+	    } else if (startswith(pathname, "word/header") ||
+		       startswith(pathname, "word/footer")) {
+		auto i = content.size();
+		total = archive_entry_size(entry);
+		content.resize(i + total);
+		size = archive_read_data(archive_obj, &content[i], total);
+
+		if (size > 0) {
+		    content.resize(i + size);
+		} else {
+		    // Ignore this as header/footer may not be present
+		    content.resize(i);
+		}
+	    } else if (pathname == "docProps/core.xml") {
+		// docProps/core.xml stores meta data
+		total = archive_entry_size(entry);
+		string metadata(total, ' ');
+		size = archive_read_data(archive_obj, &metadata[0], total);
+		if (size > 0) {
+		    metadata.resize(size);
+		    parse_metadata(metadata, title, keywords, author);
+		}
+	    }
+	}
+    } else if (startswith(tail, "spreadsheetml.")) {
+	string shared_strings, sheets;
+	while (archive_read_next_header(archive_obj, &entry) == ARCHIVE_OK) {
+	    string pathname = archive_entry_pathname(entry);
+	    if (pathname == "xl/styles.xml" ||
+		pathname == "xl/workbook.xml" ||
+		pathname == "xl/sharedStrings.xml") {
+		auto i = shared_strings.size();
+		total = archive_entry_size(entry);
+		shared_strings.resize(i + total);
+		size = archive_read_data(archive_obj, &shared_strings[i],
+					 total);
+
+		if (size > 0)
+		    shared_strings.resize(i + size);
+		else
+		    shared_strings.resize(i);
+	    } else if (startswith(pathname, "xl/worksheets/sheet")) {
+		auto i = sheets.size();
+		total = archive_entry_size(entry);
+		sheets.resize(i + total);
+		size = archive_read_data(archive_obj, &sheets[i], total);
+
+		if (size <= 0) {
+		    error = "Libarchive was not able to extract data from " +
+			    pathname;
+		    return false;
+		}
+		sheets.resize(i + size);
+	    } else if (pathname == "docProps/core.xml") {
+		total = archive_entry_size(entry);
+		string metadata(total, ' ');
```

Comment:
The content we fill the string to have here shouldn't matter as we resize to remove any excess we didn't read into, but it'd be more consistent to use the same fill byte in cases like this as we get for `sheets.resize(i + total);` which would be a zero byte (i.e. `'\0'`).  Zero bytes should also make it more obvious to the person debugging if we did end up leaving some of this fill in the read data, whereas spaces will probably look like valid data.

Reply:
I have changed this.